### DISCUSSION TITLE
Global Hotkey Solution for v11 in windows?

### DISCUSSION THREAD

**Discussion by butaixianran** [AvaloniaUI/Avalonia#14413] [General]
There was some solutions for global hotkey for older version of avalonia, for example:  
https://github.com/AvaloniaUI/Avalonia/discussions/8823  
or   
https://github.com/8/GlobalHotKeys/blob/master/src/Examples/Avalonia/App.axaml.cs  

**But they all need some API that v11 doesn't have anymore.**  

Take the package `GlobalHotKeys.Windows` as example, which only need a scheduler to work. Avalonia had a `Avalonia.Threading.AvaloniaScheduler.Instance` for that, but it is not in v11 anymore (but still in API document).

So, is there any solution for Global hotkey now?

Even I can get a handle with   
`TopLevel.GetTopLevel(desktop.MainWindow).TryGetPlatformHandle().Handle;`  
I still need a way to hook app to window's message system to response hotkey. But seems there is no API can do that with v11.  

 Need a solution here, thanks

**Comment by timunie** [comment]
I am afraid you need to roll your own solution here which uses some Win32 API

**Comment by maxkatz6** [comment]
> I still need a way to hook app to window's message system to response hotkey. But seems there is no API can do that with v11.

You can with Win32 hook APIs https://github.com/AvaloniaUI/Avalonia/discussions/8823
Or in 11.1 (currently nightly builds) there is a helper class for [Win32](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/Platform/Win32SpecificOptions.cs)

**Reply by maxkatz6** [reply]
But I don't think you will have a crossplatform solution easily.

**Reply by thevortexcloud** [reply]
There is this library: https://github.com/TolikPylypchuk/SharpHook

But yeah, writing your own is typically not too hard. Although you do often run into security problems on things that are not Windows. From memory SharpHook is dependent on X11 APIs so it probably does not work too well on most modern Linux distros.

**Reply by maxkatz6** [reply]
Haven't used, but sounds like a great solution for butaixianran

**Reply by thevortexcloud** [reply]
I have not used it either, but I have seen a lot of people suggest it and it has decent amounts of stars on github.

**Comment by butaixianran** [comment]
v11.1's win32 hook api looks nice. 
But I'm gonna use `https://github.com/TolikPylypchuk/SharpHook` for now until it is stable.  

A problem of `SharpHook` is, it listens on all key and mouse events, so it is also running on another thread which can not update UI. I have to do some work round with win32 API for that. That's not the best way for global hotkey. But it is a simple solution for now.

**Reply by thevortexcloud** [reply]
> so it is also running on another thread which can not update UI

You should be able to just use `Dispatcher` to put it back onto the UI thread?
